File: app/src/main/java/io/github/ovso/blackbox/data/network/SearchRequest.java
package io.github.ovso.blackbox.data.network;

import android.text.TextUtils;
import io.github.ovso.blackbox.Security;
import io.github.ovso.blackbox.data.KeyName;
import io.github.ovso.blackbox.data.network.model.Search;
import io.reactivex.Single;
import java.util.HashMap;
import java.util.Map;
import javax.inject.Inject;
import okhttp3.Headers;

public class SearchRequest extends BaseRequest<SearchService> {

  @Inject public SearchRequest() {

  }

  @Override protected Class<SearchService> getApiClass() {
    return SearchService.class;
  }

  @Override protected Headers createHeaders() {

    return new Headers.Builder().build();
  }

  @Override protected String getBaseUrl() {
    return ApiEndPoint.SEARCH.getUrl();
  }

  public Single<Search> getResult(String q, String pageToken) {
    return getApi().getResult(createQueryMap(q, pageToken));
  }

  private Map<String, Object> createQueryMap(String q, String pageToken) {
    Map<String, Object> queryMap = new HashMap<>();
    queryMap.put(KeyName.Q.get(), q);
    queryMap.put(KeyName.MAX_RESULTS.get(), 50);
    queryMap.put(KeyName.ORDER.get(), "viewCount");
    queryMap.put(KeyName.TYPE.get(), "video");
    queryMap.put(KeyName.VIDEO_SYNDICATED.get(), "any");
    queryMap.put(KeyName.KEY.get(), Security.KEY.getValue());
    queryMap.put(KeyName.PART.get(), "snippet");
    if (!TextUtils.isEmpty(pageToken)) {
      queryMap.put(KeyName.PAGE_TOKEN.get(), pageToken);
    }
    return queryMap;
  }
}

File: app/src/main/java/io/github/ovso/blackbox/data/network/SearchRequest.kt
package io.github.ovso.blackbox.data.network

import android.text.TextUtils
import io.github.ovso.blackbox.data.KeyName
import io.github.ovso.blackbox.data.Keys
import io.github.ovso.blackbox.data.network.model.Search
import io.reactivex.rxjava3.core.Single
import okhttp3.Headers
import java.util.*
import javax.inject.Inject

class SearchRequest @Inject constructor() : BaseRequest<SearchService>() {
  override val apiClass: Class<SearchService>
    get() = SearchService::class.java
  override val baseUrl: String
    get() = ApiEndPoint.SEARCH.url

  override fun createHeaders() = Headers.Builder().build()

  fun getResult(
    q: String,
    pageToken: String?
  ): Single<Search> {
    return api.getResult(createQueryMap(q, pageToken))
  }

  private fun createQueryMap(
    q: String,
    pageToken: String?
  ): Map<String, Any> {
    val queryMap = HashMap<String, Any>()
    queryMap[KeyName.Q.get()] = q
    queryMap[KeyName.MAX_RESULTS.get()] = 50
    queryMap[KeyName.ORDER.get()] = "date"
    queryMap[KeyName.TYPE.get()] = "video"
    queryMap[KeyName.VIDEO_SYNDICATED.get()] = "true"
    queryMap[KeyName.KEY.get()] = Keys.KEY
    queryMap[KeyName.PART.get()] = "snippet"
    if (!TextUtils.isEmpty(pageToken)) queryMap[KeyName.PAGE_TOKEN.get()] = pageToken!!
    return queryMap
  }
}
